File: src/fuzzy.ts
/**
 * Typo tolerance for search: given a query token with no exact vocabulary
 * match, find nearby real words (by edit distance) so a misspelling like
 * "permiere" still surfaces "premiere". A BK-tree makes "find all words
 * within distance N" fast without scanning the whole vocabulary per token.
 */

/** Classic dynamic-programming Levenshtein distance. */
export function levenshtein(a: string, b: string): number {
  const m = a.length;
  const n = b.length;
  if (m === 0) return n;
  if (n === 0) return m;

  let prev = new Array(n + 1);
  let curr = new Array(n + 1);
  for (let j = 0; j <= n; j++) prev[j] = j;

  for (let i = 1; i <= m; i++) {
    curr[0] = i;
    for (let j = 1; j <= n; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min(
        prev[j] + 1, // deletion
        curr[j - 1] + 1, // insertion
        prev[j - 1] + cost // substitution
      );
    }
    [prev, curr] = [curr, prev];
  }
  return prev[n];
}

/** How permissive fuzzy matching is, scaled to word length -- matches
 * Elasticsearch's `fuzziness: AUTO` default. Short words are exact-only
 * (edit distance 1 on a 3-letter word can mean almost any other 3-letter
 * word); longer words can tolerate more typos before becoming ambiguous. */
export function maxEditDistance(tokenLength: number): number {
  if (tokenLength <= 3) return 0;
  if (tokenLength <= 5) return 1;
  return 2;
}

interface BKNode {
  word: string;
  children: Map<number, BKNode>;
}

/** BK-tree: a metric tree over edit distance. Insert is O(depth) per word;
 * querying "all words within distance N" prunes almost the whole tree via
 * the triangle inequality instead of comparing against every vocabulary
 * word. Built once per corpus, queried once per query token. */
export class BKTree {
  private root: BKNode | undefined;

  insert(word: string): void {
    if (!this.root) {
      this.root = { word, children: new Map() };
      return;
    }
    let node = this.root;
    for (;;) {
      const dist = levenshtein(word, node.word);
      if (dist === 0) return; // already present
      const child = node.children.get(dist);
      if (!child) {
        node.children.set(dist, { word, children: new Map() });
        return;
      }
      node = child;
    }
  }

  /** All vocabulary words within `maxDist` edits of `word`, closest first. */
  search(word: string, maxDist: number): string[] {
    if (!this.root || maxDist <= 0) return [];
    const results: { word: string; dist: number }[] = [];
    const stack: BKNode[] = [this.root];
    while (stack.length > 0) {
      const node = stack.pop()!;
      const dist = levenshtein(word, node.word);
      if (dist > 0 && dist <= maxDist) results.push({ word: node.word, dist });
      // triangle inequality: only children whose edge-distance is within
      // [dist - maxDist, dist + maxDist] can possibly be close enough
      for (const [edgeDist, child] of node.children) {
        if (edgeDist >= dist - maxDist && edgeDist <= dist + maxDist) stack.push(child);
      }
    }
    results.sort((a, b) => a.dist - b.dist);
    return results.map((r) => r.word);
  }
}

/**
 * Prefix matching -- a different problem from typo tolerance, and one edit
 * distance can't solve: a short query term like "ux" and its intended full
 * word "uxp" often differ by more edits than the query term is even long
 * (irrelevant here, but a query of "auth" against "authentication" differs
 * by 10 edits -- there's no reasonable edit-distance threshold that catches
 * both real typos and genuine partial-word searches without also matching
 * everything). This is the classic "search-as-you-type" case: the query is
 * a deliberate prefix of a longer real word, not a misspelling of it.
 * A sorted array + binary search finds the whole prefix range in
 * O(log n + k) without scanning the vocabulary.
 */
export class PrefixIndex {
  private sorted: string[];

  constructor(words: Iterable<string>) {
    this.sorted = [...new Set(words)].sort();
  }

  /** All vocabulary words starting with `prefix` (excluding an exact match,
   * which is handled separately), shortest -- closest to what was typed --
   * first. */
  search(prefix: string, maxResults = 5): string[] {
    if (!prefix) return [];
    const n = this.sorted.length;
    let lo = 0;
    let hi = n;
    while (lo < hi) {
      const mid = (lo + hi) >>> 1;
      if (this.sorted[mid] < prefix) lo = mid + 1;
      else hi = mid;
    }
    const results: string[] = [];
    for (let i = lo; i < n; i++) {
      if (!this.sorted[i].startsWith(prefix)) break;
      if (this.sorted[i] !== prefix) results.push(this.sorted[i]);
    }
    results.sort((a, b) => a.length - b.length);
    return results.slice(0, maxResults);
  }
}

export function buildVocabularyIndex(tokenLists: string[][]): BKTree {
  const tree = new BKTree();
  const seen = new Set<string>();
  for (const tokens of tokenLists) {
    for (const t of tokens) {
      if (!seen.has(t)) {
        seen.add(t);
        tree.insert(t);
      }
    }
  }
  return tree;
}

export function buildPrefixIndex(tokenLists: string[][]): PrefixIndex {
  const words = new Set<string>();
  for (const tokens of tokenLists) for (const t of tokens) words.add(t);
  return new PrefixIndex(words);
}

File: src/fuzzy.rs
//! Port of src/fuzzy.ts -- typo tolerance for search via a BK-tree (edit
//! distance) plus a sorted-array prefix index, faithful literal port.

use std::collections::HashMap;

/// Classic dynamic-programming Levenshtein distance.
pub fn levenshtein(a: &str, b: &str) -> usize {
    let a: Vec<char> = a.chars().collect();
    let b: Vec<char> = b.chars().collect();
    let m = a.len();
    let n = b.len();
    if m == 0 {
        return n;
    }
    if n == 0 {
        return m;
    }
    let mut prev: Vec<usize> = (0..=n).collect();
    let mut curr = vec![0usize; n + 1];
    for i in 1..=m {
        curr[0] = i;
        for j in 1..=n {
            let cost = if a[i - 1] == b[j - 1] { 0 } else { 1 };
            curr[j] = (prev[j] + 1).min(curr[j - 1] + 1).min(prev[j - 1] + cost);
        }
        std::mem::swap(&mut prev, &mut curr);
    }
    prev[n]
}

/// How permissive fuzzy matching is, scaled to word length -- matches
/// Elasticsearch's `fuzziness: AUTO` default. Short words are exact-only;
/// longer words can tolerate more typos before becoming ambiguous.
pub fn max_edit_distance(token_length: usize) -> usize {
    if token_length <= 3 {
        0
    } else if token_length <= 5 {
        1
    } else {
        2
    }
}

struct BkNode {
    word: String,
    children: HashMap<usize, BkNode>,
}

/// BK-tree: a metric tree over edit distance. Insert is O(depth) per word;
/// querying "all words within distance N" prunes almost the whole tree via
/// the triangle inequality instead of comparing against every vocabulary
/// word. Built once per corpus, queried once per query token.
pub struct BKTree {
    root: Option<BkNode>,
}

impl BKTree {
    pub fn new() -> Self {
        Self { root: None }
    }

    pub fn insert(&mut self, word: &str) {
        let Some(root) = &mut self.root else {
            self.root = Some(BkNode { word: word.to_string(), children: HashMap::new() });
            return;
        };
        let mut node = root;
        loop {
            let dist = levenshtein(word, &node.word);
            if dist == 0 {
                return; // already present
            }
            if node.children.contains_key(&dist) {
                node = node.children.get_mut(&dist).unwrap();
            } else {
                node.children.insert(dist, BkNode { word: word.to_string(), children: HashMap::new() });
                return;
            }
        }
    }

    /// All vocabulary words within `max_dist` edits of `word`, closest first.
    pub fn search(&self, word: &str, max_dist: usize) -> Vec<String> {
        let Some(root) = &self.root else { return Vec::new() };
        if max_dist == 0 {
            return Vec::new();
        }
        let mut results: Vec<(String, usize)> = Vec::new();
        let mut stack: Vec<&BkNode> = vec![root];
        while let Some(node) = stack.pop() {
            let dist = levenshtein(word, &node.word);
            if dist > 0 && dist <= max_dist {
                results.push((node.word.clone(), dist));
            }
            // triangle inequality: only children whose edge-distance is
            // within [dist - maxDist, dist + maxDist] can possibly be close
            // enough
            for (&edge_dist, child) in &node.children {
                let lower = dist.saturating_sub(max_dist);
                if edge_dist >= lower && edge_dist <= dist + max_dist {
                    stack.push(child);
                }
            }
        }
        results.sort_by_key(|(_, d)| *d);
        results.into_iter().map(|(w, _)| w).collect()
    }
}

impl Default for BKTree {
    fn default() -> Self {
        Self::new()
    }
}

/// Prefix matching -- a different problem from typo tolerance, and one
/// edit distance can't solve (a query of "auth" against "authentication"
/// differs by 10 edits, but is a genuine 4-character prefix match). A
/// sorted array + binary search finds the whole prefix range in
/// O(log n + k) without scanning the vocabulary.
pub struct PrefixIndex {
    sorted: Vec<String>,
}

impl PrefixIndex {
    pub fn new(words: impl IntoIterator<Item = String>) -> Self {
        let mut set: std::collections::BTreeSet<String> = words.into_iter().collect();
        let sorted: Vec<String> = std::mem::take(&mut set).into_iter().collect();
        Self { sorted }
    }

    /// All vocabulary words starting with `prefix` (excluding an exact
    /// match, which is handled separately), shortest -- closest to what was
    /// typed -- first.
    pub fn search(&self, prefix: &str, max_results: usize) -> Vec<String> {
        if prefix.is_empty() {
            return Vec::new();
        }
        let n = self.sorted.len();
        let mut lo = 0usize;
        let mut hi = n;
        while lo < hi {
            let mid = (lo + hi) / 2;
            if self.sorted[mid].as_str() < prefix {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        let mut results: Vec<String> = Vec::new();
        for item in &self.sorted[lo..] {
            if !item.starts_with(prefix) {
                break;
            }
            if item != prefix {
                results.push(item.clone());
            }
        }
        results.sort_by_key(|s| s.chars().count());
        results.truncate(max_results);
        results
    }
}

pub fn build_vocabulary_index(token_lists: &[Vec<String>]) -> BKTree {
    let mut tree = BKTree::new();
    let mut seen = std::collections::HashSet::new();
    for tokens in token_lists {
        for t in tokens {
            if seen.insert(t.clone()) {
                tree.insert(t);
            }
        }
    }
    tree
}

pub fn build_prefix_index(token_lists: &[Vec<String>]) -> PrefixIndex {
    let mut words = std::collections::HashSet::new();
    for tokens in token_lists {
        for t in tokens {
            words.insert(t.clone());
        }
    }
    PrefixIndex::new(words)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn levenshtein_basic() {
        assert_eq!(levenshtein("kitten", "sitting"), 3);
        assert_eq!(levenshtein("", "abc"), 3);
        assert_eq!(levenshtein("abc", "abc"), 0);
    }

    #[test]
    fn bktree_finds_typo() {
        let mut tree = BKTree::new();
        for w in ["premiere", "project", "product"] {
            tree.insert(w);
        }
        let results = tree.search("permiere", max_edit_distance("permiere".len()));
        assert!(results.contains(&"premiere".to_string()));
    }

    #[test]
    fn prefix_index_finds_shortest_first() {
        let idx = PrefixIndex::new(vec!["uxp".to_string(), "uxplugin".to_string(), "unrelated".to_string()]);
        let results = idx.search("ux", 5);
        assert_eq!(results, vec!["uxp".to_string(), "uxplugin".to_string()]);
    }
}
